toolchain/check/facet_type.cpp

```diff
@@ -196,34 +194,40 @@ auto InitialFacetTypeImplWitness(
       // forming the facet type because the type of the associated constant
       // was symbolic.
       auto converted_inst_id = ConvertToValueOfType(
-          context, context.insts().GetLocId(facet_type_inst_id),
-          context.constant_values().GetInstId(rewrite_value),
+          context, context.insts().GetLocId(facet_type_inst_id), rewrite_value,
           assoc_const_type_id);
-      rewrite_value = context.constant_values().Get(converted_inst_id);
+      // Canonicalize the converted constant value.
+      converted_inst_id =
+          context.constant_values().GetConstantInstId(converted_inst_id);
       // The result of conversion can be non-constant even if the original
       // value was constant.
-      if (!rewrite_value.is_constant() &&
-          rewrite_value != SemIR::ErrorInst::SingletonConstantId) {
+      if (converted_inst_id.has_value()) {
+        rewrite_value = converted_inst_id;
+      } else if (rewrite_value != SemIR::ErrorInst::SingletonInstId) {
         const auto& assoc_const = context.associated_constants().Get(
             assoc_constant_decl->assoc_const_id);
         CARBON_DIAGNOSTIC(
             AssociatedConstantNotConstantAfterConversion, Error,
-            "associated constant {0} given value that is not constant "
-            "after conversion to {1}",
-            SemIR::NameId, SemIR::TypeId);
-        context.emitter().Emit(facet_type_inst_id,
-                               AssociatedConstantNotConstantAfterConversion,
-                               assoc_const.name_id, assoc_const_type_id);
-        rewrite_value = SemIR::ErrorInst::SingletonConstantId;
+            "associated constant {0} given value {1} that is not constant "
+            "after conversion to {2}",
+            SemIR::NameId, InstIdAsType, SemIR::TypeId);
+        context.emitter().Emit(
+            facet_type_inst_id, AssociatedConstantNotConstantAfterConversion,
+            assoc_const.name_id, rewrite_value, assoc_const_type_id);
+        rewrite_value = SemIR::ErrorInst::SingletonInstId;
       }
     }
 
-    auto rewrite_inst_id = context.constant_values().GetInstId(rewrite_value);
+    CARBON_CHECK(rewrite_value ==
+                     context.constant_values().GetConstantInstId(rewrite_value),
+                 "Rewritten value for associated constant is not canonical.");
+
+    auto rewrite_inst_id = rewrite_value;
     table_entry = AddInst<SemIR::ImplWitnessAssociatedConstant>(
         context, witness_loc_id,
         {.type_id = context.insts().Get(rewrite_inst_id).type_id(),
          .inst_id = rewrite_inst_id});
-    table_entry = context.constant_values().GetInstId(rewrite_value);
+    table_entry = rewrite_value;
```

Comment:
This sequence of instructions is pretty confusing. What's going on? Why make `rewrite_inst_id` at all? Why assign `table_entry` twice?

Reply:
Good catch, the double-store to `table_entry` here was a bug that Dana just fixed in #5283 :)

Reply:
Merged that change and cleaned up the unnecessary variable.